filebeat/beater/filebeat.go

```diff
@@ -82,19 +81,27 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 		logp.Err("Could not start registrar: %v", err)
 	}
 	// Stopping registrar will write last state
-	defer registrar.Stop()
+	defer func() {
+		// Stop logger to make sure not new events are added
+		registrarLogger.Close()
+		// Stop logger itself
+		registrar.Stop()
+	}()
 
 	// Start publisher
 	publisher.Start()
 	// Stopping publisher (might potentially drop items)
-	defer publisher.Stop()
-	defer successLogger.Close()
+	defer func() {
+		// Closing channel to publisher
+		publisherChan.Close()
+		// Stopping publisher itself
+		publisher.Stop()
+	}()
```

Comment:
closing `publisherChan` after spooler can make the spooler hang if:
- publisher is blocked by output
- channel being full
- spooler is trying to push via publisherChan

=> shutdown hangs